src/utils/bigint.ts

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { IEncoding } from 'level-transcoder';
+import { z } from 'zod';
+
+/**
+ * An object equivalent to the native global JSON, providing `parse()` and `stringify()` functions with compatible signatures, except
+ * for the `reviver` and `replacer` parameters that are not supported to prevent accidental override of the custom BigInt behavior.
+ *
+ * If the JSON string to be parsed contains large integers that would lose precision with the `number` type, they're parsed as `bigint`s,
+ * and analogously for `stringify`. The `any` type is allowed as it conforms to the original signatures.
+ */
+export const JSONBigInt = {
+  /* eslint-disable @typescript-eslint/no-explicit-any */
+  parse(text: string): any {
+    function bigIntReviver(_key: string, value: any, context: { source: string }): any {
+      if (!Number.isInteger(value)) {
+        // No special handling needed for non-integer values.
+        return value;
+      }
+
+      let { source } = context;
+      if (source.includes('e') || source.includes('E')) {
+        // We explicitly prohibit JSONs with exponential notation (such as 10e2) as they cannot be parsed to BigInt.
+        throw Error(`exponential notation is not supported in "${text}"`);
+      }
+
+      if (source.includes('.')) {
+        // If value is an integer and contains a '.', it must be like '123.0', so we extract the integer part only.
+        let zeroes: string;
+        [source, zeroes] = source.split('.');
+
+        if (zeroes.split('').some(char => char !== '0')) {
+          // This case shouldn't happen but we'll prohibit it to be safe. For example, if the source is
+          // '12345678901234567890.1', JS will parse it as an integer with loss of precision, `12345678901234567000`.
+          throw Error(`large float will lose precision! in "${text}"`);
+        }
+      }
+
+      const bigIntValue = BigInt(source);
+      if (bigIntValue !== BigInt(value)) {
+        // If the parsed value is an integer and its BigInt representation is a different value,
+        // it means we lost precision, so we return the BigInt.
+        return bigIntValue;
```

Comment:
> Why don't we always return a `BigInt`?

Because this will be called for schemas/functions that expect `Number` too, not just `BigInt`.

> How are we going to handle operations if this is sometimes int and sometimes BigInt?

The way we differentiate between them is through Zod, which is exactly the reason why I had to use it. In other words, the custom JSON parser will only return a `BigInt` if the value is large enough, and Zod will take care of validating the correct types — for example, converting a small `Number` such as `123` to a `123n` when necessary.